frontend/catalyst/jax_tracer.py

```diff
@@ -689,45 +694,137 @@ def trace_quantum_function(
+
+            # I don't think we can now if a transform is a batch transform until we return more than 1 tape.
+            invalid_state = is_program_transformed and not are_batch_transforms_valid
+            # TODO: Consider whether issuing a warning, or an exception is better.
+            assert not invalid_state
+
+            del quantum_tape
         # (2) - Quantum tracing
+        results = []
+        results_tracers = []
+        results_abstract = []
+        for tape in tapes:
+            with EvaluationContext.frame_tracing_context(ctx, trace):
+                qdevice_p.bind(spec="kwargs", val=str(device.backend_kwargs))
+                qdevice_p.bind(spec="backend", val=device.backend_name)
+                qreg_in = qalloc_p.bind(len(device.wires))
+                qrp_out = trace_quantum_tape(tape, device, qreg_in, ctx, trace)
+                _, pytree_measurements = tree_flatten(tape.measurements)
+                out_classical_tracers, out_classical_tree = trace_quantum_measurements(
+                    device,
+                    qrp_out,
+                    # If the program is batched, that means that it was transformed.
+                    # If it was transformed, that means that the program might have
+                    # changed the output. See `split_non_commuting`
+                    out_classical_tracers_or_measurements if not is_program_transformed else tape.measurements,
+                    out_tree_promise() if not is_program_transformed else pytree_measurements,
+                )
+                out_quantum_tracers = [qrp_out.actualize()]
+                qdealloc_p.bind(qreg_in)
+
+                out_classical_tracers = [trace.full_raise(t) for t in out_classical_tracers]
+                results_tracers += out_classical_tracers
+
+                jaxpr, out_type, consts = ctx.frames[trace].to_jaxpr2(
+                    out_classical_tracers + out_quantum_tracers
+                )
+                jaxpr._outvars = jaxpr._outvars[:-1]  # pylint: disable=protected-access
+                out_type = out_type[:-1]
+
+                out_avals, _ = unzip2(out_type)
+                abstract_results = tree_unflatten(
+                    out_classical_tree if not is_program_transformed else pytree_measurements,
+                    [ShapeDtypeStruct(a.shape, a.dtype, a.named_shape) for a in out_avals],
+                )
+                if is_program_transformed and len(abstract_results) == 1:
+                    results_abstract.append(abstract_results[0])
+                elif is_program_transformed:
+                    results_abstract.append(tuple(abstract_results))
+                else:
+                    results_abstract.append(abstract_results)
+                # TODO: `check_jaxpr` complains about the `AbstractQreg` type. Consider fixing.
+                # check_jaxpr(jaxpr)
+
         with EvaluationContext.frame_tracing_context(ctx, trace):
```

Comment:
Is this tracing for the post_processing function? I would add a small comment here.

Reply:
Separated post_processing tracing into its own function: https://github.com/PennyLaneAI/catalyst/pull/280/commits/656eb33d1d5528fec2bb4e9bfaa993dccd9536d3